File: API/Entities/Order.cs
namespace API.Entities;

public class Order
{
    public int Id { get; set; }

    public int AppUserId { get; set; }
    public AppUser AppUser { get; set; } = null!;

    public string OrderNumber { get; set; } = string.Empty;

    public string OrderStatus { get; set; } = "PendingPayment";
    public string PaymentStatus { get; set; } = "PendingPayment";

    public string FirstName { get; set; } = string.Empty;
    public string LastName { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public string PhoneNumber { get; set; } = string.Empty;

    public string AddressLine1 { get; set; } = string.Empty;
    public string AddressLine2 { get; set; } = string.Empty;
    public string Suburb { get; set; } = string.Empty;
    public string City { get; set; } = string.Empty;
    public string Province { get; set; } = string.Empty;
    public string PostalCode { get; set; } = string.Empty;

    public string DeliveryMethod { get; set; } = "Standard Delivery";
    public string CustomerNote { get; set; } = string.Empty;

    public decimal SubtotalAmount { get; set; }
    public decimal DeliveryFee { get; set; }
    public decimal TotalAmount { get; set; }

    public DateTime? StockReservedAt { get; set; }
    public DateTime? StockRestoredAt { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? PaidAt { get; set; }
    public DateTime? CancelledAt { get; set; }
    public DateTime? FailedAt { get; set; }

    public ICollection<OrderItem> Items { get; set; } = new List<OrderItem>();
    public ICollection<OrderPayment> Payments { get; set; } = new List<OrderPayment>();
}

File: client/src/app/_models/order.ts
export interface CreateOrder {
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;

  addressLine1: string;
  addressLine2: string;
  suburb: string;
  city: string;
  province: string;
  postalCode: string;

  deliveryMethod: string;
  customerNote: string;
}

export interface OrderItem {
  id: number;

  productId: number;
  productVariantId: number;

  productName: string;
  category: string;
  size: string;
  colour: string;
  sku: string;

  imageUrl: string;
  imageAlt: string;

  unitPrice: number;
  unitPriceText: string;

  quantity: number;

  lineTotal: number;
  lineTotalText: string;
}

export interface OrderPayment {
  id: number;

  provider: string;
  status: string;

  amount: number;
  amountText: string;

  merchantReference: string;
  gatewayPaymentId: string;

  createdAt: string;
  paidAt?: string | null;
  failedAt?: string | null;
  cancelledAt?: string | null;
}

export interface Order {
  id: number;

  orderNumber: string;

  orderStatus: string;
  paymentStatus: string;

  firstName: string;
  lastName: string;
  fullName: string;

  email: string;
  phoneNumber: string;

  addressLine1: string;
  addressLine2: string;
  suburb: string;
  city: string;
  province: string;
  postalCode: string;

  deliveryMethod: string;
  customerNote: string;

  subtotalAmount: number;
  subtotalText: string;

  deliveryFee: number;
  deliveryFeeText: string;

  totalAmount: number;
  totalText: string;

  totalItems: number;

  createdAt: string;
  updatedAt: string;
  paidAt?: string | null;
  cancelledAt?: string | null;
  failedAt?: string | null;

  items: OrderItem[];
  payments: OrderPayment[];
}

export interface OrderParams {
  search?: string;
  orderStatus?: string;
  paymentStatus?: string;
}

export interface UpdateOrderStatus {
  orderStatus: string;
}